### ISSUE TITLE
Memory Leakages in the VRAM while saving the model checkpoint for Gemma4 E4b and 26A4B.

### ISSUE THREAD

**Issue by aman-edge** [axolotl-ai-cloud/axolotl#3596]
### Please check that this issue hasn't been reported before.

- [x] I searched previous [Bug Reports](https://github.com/axolotl-ai-cloud/axolotl/labels/bug) didn't find any similar reports.

### Expected Behavior

VRAM usage after a checkpoint save should return to approximately the same level as before the save began. The checkpoint writing process should fully release any temporary memory buffers it allocates, leaving training able to continue without cumulative memory growth across successive checkpoints.
Starting memory was ~49,258 MiB. After each checkpoint save, usage should return close to that baseline rather than growing by ~12,000 MiB per save (~61,558 → ~76,124 → OOM).

### Current behaviour

Trying to train a E4B and as the model is saved it takes up some space and is never freed. `fsdp_state_dict_type: SHARDED_STATE_DICT` worked but this means I have to merge it later. 

Initially without the `fsdp_state_dict_type: SHARDED_STATE_DICT`
wandb: WARNING Saving files without folders. If you want to preserve subdirectories pass base_path to wandb.save, i.e. wandb.save("/mnt/folder/file.h5", base_path="/mnt")
wandb: WARNING Symlinked 1 file into the W&B run directory; call wandb.save again to sync new files.

Memory per gpu is approx 49258MiB /  81559MiB 

Then 
  0%|          | 32/33000 [03:40<51:10:19,  5.59s/it]
  0%|          | 33/33000 [03:44<48:00:38,  5.24s/it]

Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

Writing model shards: 100%|██████████| 1/1 [00:24<00:00, 24.31s/it]
Writing model shards: 100%|██████████| 1/1 [00:24<00:00, 24.31s/it]

  0%|          | 34/33000 [05:26<313:01:21, 34.18s/it]
  0%|          | 35/33000 [05:33<238:47:01, 26.08s/it]

Memory per gpu is approx  61558MiB /  81559MiB

  0%|          | 65/33000 [08:10<46:15:04,  5.06s/it]
  0%|          | 66/33000 [08:16<48:22:32,  5.29s/it]

Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

Writing model shards: 100%|██████████| 1/1 [00:23<00:00, 23.44s/it]
Writing model shards: 100%|██████████| 1/1 [00:23<00:00, 23.44s/it]

Memory per gpu is  76124MiB /  81559MiB

And ofc 
0%|          | 98/33000 [12:30<45:42:14,  5.00s/it]
  0%|          | 99/33000 [12:37<51:41:10,  5.66s/it]

Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

Writing model shards: 100%|██████████| 1/1 [00:21<00:00, 21.13s/it]
Writing model shards: 100%|██████████| 1/1 [00:21<00:00, 21.13s/it]

Out of memory 

Okay with `fsdp_state_dict_type: SHARDED_STATE_DICT`

FSDP Config 
```
fsdp_config:
  enable_cpu_offload: false
  fsdp_activation_checkpointing: true
  fsdp_auto_wrap_policy: TRANSFORMER_BASED_WRAP
  fsdp_cpu_ram_efficient_loading: true
  fsdp_limit_all_gathers: true
  fsdp_offload_params: false
  fsdp_reshard_after_forward: true
  fsdp_sharding_strategy: FULL_SHARD
  fsdp_state_dict_type: SHARDED_STATE_DICT  # was FULL_STATE_DICT #FULL_STATE_DICT
  fsdp_sync_module_states: true
  fsdp_transformer_layer_cls_to_wrap: Gemma4TextDecoderLayer
  fsdp_version: 2

```

  0%|          | 75/16500 [11:42<39:15:00,  8.60s/it]
  0%|          | 76/16500 [11:51<39:31:12,  8.66s/it]
  0%|          | 77/16500 [11:58<37:26:21,  8.21s/it]
  0%|          | 78/16500 [12:05<35:58:15,  7.89s/it]
  0%|          | 79/16500 [12:16<40:13:32,  8.82s/it]
  0%|          | 80/16500 [12:23<37:56:03,  8.32s/it]
                                                     

  0%|          | 80/16500 [12:23<37:56:03,  8.32s/it]
  0%|          | 81/16500 [12:30<36:19:10,  7

  51678MiB /  81559MiB 	

 1%|          | 90/16500 [13:54<39:13:55,  8.61s/it]
  1%|          | 91/16500 [14:03<40:22:43,  8.86s/it]
  1%|          | 92/16500 [14:11<38:01:36,  8.34s/it]
  1%|          | 93/16500 [14:18<37:23:27,  8.20s/it]
  1%|          | 94/16500 [14:26<37:07:36,  8.15s/it]
  1%|          | 95/16500 [14:34<35:44:28,  7.84s/it]

51678MiB /  81559MiB |

without FSDP it doesn't work and with 26A4B model the same thing happens but on the first save checkpoint. 

Starting memory was ~49,258 MiB. After each checkpoint save, usage should return close to that baseline rather than growing by ~12,000 MiB per save (~61,558 → ~76,124 → OOM).

### Steps to reproduce

1. Hardware: [X] x H100/A100 80GB GPUs
2. Model: Gemma 4 E4B (or 26A4B)
3. FSDP config:
   - fsdp_state_dict_type: FULL_STATE_DICT (or unset/default)
   - [paste your relevant fsdp config block here]
4. Set a frequent checkpoint save interval (e.g. every 32 steps)
5. Begin training and monitor VRAM with `nvidia-smi`
6. Observe VRAM after first checkpoint save (~step 33)
7. Observe VRAM after second checkpoint save (~step 65)
8. Observe VRAM after third checkpoint save (~step 98) → OOM

### Config yaml

```yaml
# Gemma 4 26B-A4B MoE QLoRA with ScatterMoE kernels
#
# Validated: 50 steps on FineTome-100k, loss 8.8 -> 1.8, single RTX 5090 (32GB)
# torch_compile=true: 21 GiB peak VRAM, ~230 tok/s, 336s total
#
# Key notes:
# - Max sequence length on 32GB GPU: 2048 (micro_batch_size=1, SDP attention).
#   4096 seq_len OOMs due to head_dim=512 math SDP materializing full score matrix.
#   Use 48GB+ GPUs for longer sequences or multi-GPU with FSDP.

base_model: google/gemma-4-E4B-it

plugins:
  - axolotl.integrations.cut_cross_entropy.CutCrossEntropyPlugin
  - axolotl.integrations.kernels.KernelsPlugin
  - axolotl.integrations.liger.LigerPlugin
use_kernels: true
use_scattermoe: true
experts_implementation: scattermoe
torch_compile: false #true
liger_layer_norm: true
liger_rope: true
liger_rms_norm: true
liger_glu_activation: true
liger_rms_norm_gated: true
strict: false

#from gemma 3 config
tokenizer_num_proc: 64
dataset_processes: 64
save_safetensors: true
auto_resume_from_checkpoints: false

chat_template: gemma4
datasets:
  - path: <>
    revision: <>
    type: chat_template

# val_set_size: 0.0005
output_dir: 

sequence_len: 2048
sample_packing: true
# eval_sample_packing: true
remove_unused_columns: true
pad_to_sequence_len: true

# for flex attension
# torch_compile: true
flex_attention: false #true
sdp_attention: true 
# flex_attention: true
#lora
# load_in_4bit: true
# quantize_moe_experts: true
# adapter: qlora
# lora_r: 16
# lora_alpha: 32
# lora_dropout: 0

# Restrict LoRA to text backbone only (skip vision/audio encoders)
# using regex to match only the text decoder attention projections.
# lora_target_modules: 'model.language_model.layers.[\d]+.(_checkpoint_wrapped_module.)?(mlp|self_attn).(up|down|gate|q|k|v|o)_proj'

# # MoE expert LoRA (3D Parameter tensors, not nn.Linear)
# lora_target_parameters:
#   - experts.gate_up_proj
#   - experts.down_proj

# lora_mlp_kernel: false
# lora_qkv_kernel: false
# lora_o_kernel: false

#for lora
# bnb_config_kwargs:
#   bnb_4bit_use_double_quant: true

# Weights & Biases configuration
use_wandb: true
wandb_project: 
wandb_entity: # Your wandb username or team name
wandb_name: 
wandb_watch: # "gradients", "all", or "false"
# wandb_log_model: "checkpoint"  # "checkpoint" to log model every save_steps, "end" to log only at end
wandb_mode: # "online", "offline", or "disabled" - leave empty for default (online)
wandb_log_model:

gradient_accumulation_steps: 4
micro_batch_size: 4
num_epochs: 5
optimizer: adamw_torch_8bit
lr_scheduler: cosine
learning_rate: 0.000001

bf16: auto
tf32: true

gradient_checkpointing: false #true #false #true
activation_offloading: false #true
# logging_steps: 1

# FA2 not supported
# sdp_attention: true

warmup_ratio: 0.1
# from gemma 4 example
# evals_per_epoch: 4
# saves_per_epoch: 1
# weight_decay: 0.0
# special_tokens:
# from gemma 3


# Logging and saving
logging_steps: 5
# save_strategy: "epoch"
# save_steps: 250
# evals_per_epoch: 1
save_total_limit: 10 # can be less 
saves_per_epoch: 100 #100 for quick testing
use_tensorboard: true
## less saves to save 
# save_strategy: "steps"
# save_steps: 4125  ## 2 nodes 5 epoch save        # 1650 / 2 = halfway
# save_total_limit: 2
# use_tensorboard: true

# Best model selection
load_best_model_at_end: true  # Load the best checkpoint at the end
# metric_for_best_model: "eval_loss"  # Use validation loss to determine best model
greater_is_better: false  # Lower loss is better
# early_stopping_patience: 3  # Stop if no improvement for 3 evaluations (optional)
early_stopping_patience:


fsdp_config:
  enable_cpu_offload: false
  fsdp_activation_checkpointing: true
  fsdp_auto_wrap_policy: TRANSFORMER_BASED_WRAP
  fsdp_cpu_ram_efficient_loading: true
  fsdp_limit_all_gathers: true
  fsdp_offload_params: false
  fsdp_reshard_after_forward: true
  fsdp_sharding_strategy: FULL_SHARD
  fsdp_state_dict_type: FULL_STATE_DICT  ## to make things work do SHARDED_STATE_DICT  # was FULL_STATE_DICT #FULL_STATE_DICT
  fsdp_sync_module_states: true
  fsdp_transformer_layer_cls_to_wrap: Gemma4TextDecoderLayer
  fsdp_version: 2

# fsdp_config:
#   enable_cpu_offload: false
#   fsdp_activation_checkpointing: true
#   fsdp_auto_wrap_policy: TRANSFORMER_BASED_WRAP
#   fsdp_cpu_ram_efficient_loading: true
#   fsdp_limit_all_gathers: true
#   fsdp_offload_params: false
#   fsdp_reshard_after_forward: true
#   fsdp_sharding_strategy: FULL_SHARD
#   fsdp_state_dict_type: SHARDED_STATE_DICT
#   fsdp_final_state_dict_type: FULL_STATE_DICT
#   fsdp_sync_module_states: true
#   fsdp_transformer_layer_cls_to_wrap: Gemma4TextDecoderLayer
#   fsdp_version: 2

# Add DDP timeout for multi-node training
ddp_timeout: 7200  # 2 hours in seconds

# Additional settings for stability
dataloader_num_workers: 64
dataloader_pin_memory: false
```

### Possible solution

Most likely culprits to mention:

Explicit torch.cuda.empty_cache() after checkpoint save — the simplest fix, worth trying. Memory may be allocated but not freed back to the OS after the save completes.

FSDP gather during save — with FULL_STATE_DICT, FSDP temporarily gathers the full model onto rank 0 (or all ranks depending on config). If the gathered tensors aren't properly discarded after saving, they'd persist in VRAM. The fix would be ensuring the gathered state dict is explicitly deleted and empty_cache() called after.
offload_to_cpu flag — in FSDP's FullStateDictConfig, setting offload_to_cpu=True and rank0_only=True can significantly reduce peak VRAM during saves by offloading the gathered weights to CPU RAM instead.

### Which Operating Systems are you using?

- [x] Linux
- [ ] macOS
- [ ] Windows

### Python Version

e Python ≥3.11

### axolotl branch-commit

git+https://github.com/axolotl-ai-cloud/ml-cross-entropy.git@fec1a888e6f4ad7e6270ea7b02186e56c76f5ac2"

### Acknowledgements

- [x] My issue title is concise, descriptive, and in title casing.
- [x] I have searched the existing issues to make sure this bug has not been reported yet.
- [x] I am using the latest version of axolotl.
- [x] I have provided enough information for the maintainers to reproduce and diagnose the issue.

**Comment by reallyticsai** [comment]
One thing we've seen in large-scale distributed training with FSDP is that memory fragmentation can cause these kinds of VRAM leaks after checkpointing. Using `fsdp_state_dict_type: SHARDED_STATE_DICT` is often the safest in terms of memory but requires manual merging later, as you noted. Another trick that helped us was explicitly calling `torch.cuda.empty_cache()` right after checkpoint saving to clear any cached GPU memory, though it doesn't always fully solve the problem. Also, double-check if your checkpoint saving code holds onto any references of the model or optimizer that might prevent PyTorch from freeing those buffers. If you’re using mixed precision or gradient checkpointing, those can sometimes exacerbate memory retention during saves. Finally, updating to the latest PyTorch/FSDP versions can help, since there have been some recent memory cleanup improvements in the state dict APIs.

**Comment by thad0ctor** [comment]
I  was OOM with Gemma4 31B about 50-70 steps into training regardless of lora, qlora, etc, even when starting out a run a 50% GPU capacity,VRAM would slowly climb before eventually OOM.  That said this may not be a saving issue and a garbage collection issue.  Try gc_steps: 10 (or another value) in your config, still get VRAM increasing with every training step about but this helps knock it back down, I sync it with my logging steps.

I haven't experience this behavior with other models in the past, it could be unique to gemma. This reportsmake me think there could be something either with the model or how it behaves with CUDA and freeing memory:

https://forums.developer.nvidia.com/t/gemma-4-vlm-vram-host-memory-leak-full-investigation-report/366221
